FS3000 sensor

kahrendt opened the pull request:
# What does this implement/fix?

This pull request adds support for the FS3000 Air Velocity Sensor. It includes support for the two available models, 1005 and 1015.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes (not applicable)

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2715

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
sensor:
  - platform: fs3000
    name: "Air Velocity"
    model: 1005
    update_interval: 60s

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kahrendt committed:
Add support for FS3000 sensor.

kahrendt committed:
add fs3000 to test yaml

kahrendt committed:
Clean up code with clang.

kahrendt committed:
Clean up sensor.py file.

kahrendt committed:
Update CODEOWNERS file.

SeByDocKy commented:
Very interesting sensor.... and thanks for the ESPhome integration.... I may use it in a big project

kahrendt committed:
Merge branch 'dev' into fs3000-sensor

kahrendt committed:
Merge branch 'dev' into fs3000-sensor

jesserockz commented on esphome/components/fs3000/sensor.py:
```diff
@@ -0,0 +1,53 @@
+    "FS3000Component", cg.PollingComponent, i2c.I2CDevice, sensor.Sensor
+)
+
+CONFIG_SCHEMA = cv.All(
```
No need to wrap in `All`
```suggestion
CONFIG_SCHEMA = (
```

jesserockz commented on esphome/components/fs3000/sensor.py:
```diff
@@ -0,0 +1,53 @@
+    )
+    .extend(
+        {
+            cv.GenerateID(): cv.declare_id(FS3000Component),
```
This is not needed as the `sensor_schema()` adds it already
```suggestion
```

jesserockz commented on esphome/components/fs3000/sensor.py:
```diff
@@ -0,0 +1,53 @@
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+    await sensor.register_sensor(var, config)
```
```suggestion
    var = await sensor.new_sensor(config)
    await cg.register_component(var, config)
    await i2c.register_i2c_device(var, config)
```

jesserockz commented on esphome/components/fs3000/fs3000.cpp:
```diff
@@ -0,0 +1,107 @@
+void FS3000Component::set_model(FS3000Model model) {
+  this->model_ = model;
+
+  if (model_ == FIVE) {
+    // datasheet gives 9 points to interpolate from for the 1005 model
+    static const uint16_t RAW_DATA_POINTS_1005[9] = {409, 915, 1522, 2066, 2523, 2908, 3256, 3572, 3686};
+    static const float MPS_DATA_POINTS_1005[9] = {0.0, 1.07, 2.01, 3.0, 3.97, 4.96, 5.98, 6.99, 7.23};
+
+    std::copy(RAW_DATA_POINTS_1005, RAW_DATA_POINTS_1005 + 9, this->raw_data_points_);
+    std::copy(MPS_DATA_POINTS_1005, MPS_DATA_POINTS_1005 + 9, this->mps_data_points_);
+  } else if (model_ == FIFTEEN) {
+    // datasheet gives 13 points to extrapolate form for the 1015 model
+    static const uint16_t RAW_DATA_POINTS_1015[13] = {409,  1203, 1597, 1908, 2187, 2400, 2629,
+                                                      2801, 3006, 3178, 3309, 3563, 3686};
+    static const float MPS_DATA_POINTS_1015[13] = {0.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 13.0, 15.0};
+
+    std::copy(RAW_DATA_POINTS_1015, RAW_DATA_POINTS_1015 + 13, this->raw_data_points_);
+    std::copy(MPS_DATA_POINTS_1015, MPS_DATA_POINTS_1015 + 13, this->mps_data_points_);
+  }
```
I would prefer this initialization is done in `setup` and leave the `setter` to be just a `setter` for the `model` variable

kahrendt replied:
Makes sense, thanks!

jesserockz commented on esphome/components/fs3000/fs3000.h:
```diff
@@ -0,0 +1,34 @@
+  void set_model(FS3000Model model);
+
+ protected:
+  FS3000Model model_{FIVE};
```
`set_model` is always called by code generation so a default is not needed in the header
```suggestion
  FS3000Model model_{};
```

jesserockz commented on esphome/components/fs3000/fs3000.cpp:
```diff
@@ -0,0 +1,107 @@
+  if (!this->read_bytes_raw(data, 5)) {
+    this->status_set_warning();
+    ESP_LOGW(TAG, "Error reading data from FS3000");
+    publish_state(NAN);
```
```suggestion
    this->publish_state(NAN);
```

jesserockz commented on esphome/components/fs3000/fs3000.cpp:
```diff
@@ -0,0 +1,107 @@
+  ESP_LOGV(TAG, "Got raw reading=%i", raw_value);
+
+  // convert and publish the raw value into m/s using the table of data points in the datasheet
+  publish_state(fit_raw_(raw_value));
```
```suggestion
  this->publish_state(fit_raw_(raw_value));
```

kahrendt committed:
Apply suggestions from code review regarding sensor.py

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kahrendt committed:
Apply suggestions from code review for basic issues regarding C++ code

- removed unnecessary default for FS3000Model
- use "this->" before any sensor update

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

kahrendt committed:
Move model setup to overall setup function.

kahrendt committed:
Remove unneeded CONF_ID from sensor.py

kahrendt committed:
Run clang-format

kahrendt commented:
Thank you for taking the time to review my pull request! This is all fairly new to me, and I appreciate your help. I made the requested changes, including the in-depth setting up code for each module into the setup() function, as well as fixed the sensor.py to remove the unnecessary components.

kahrendt committed:
Move set_model code to header file now that it is simplified

jesserockz committed:
Update fs3000.h
